### PR TITLE
fix(clustering) Handle connection error on version negotiation

### PR DESCRIPTION
**Pull request by javierguerragiraldez**
If the CP doesn't respond to the version negotiation request, the DP
will assume the "old" protocol.

A failure to establish the websocket connection repeats the negotiation
attempt, as it would probably indicate a control plane rebooting or real
connectivity issues.  Later errors would (first) attempt only reconnect,
as they would likely be transient issue.  But if that reconnection
fails, it would again retry to renegotiate.

### CODE REVIEW

**Comment by javierguerragiraldez** [comment]
Changing version negotiation flow.  Do not merge

File: kong/clustering/init.lua [hunk 1]
```diff
@@ -205,14 +205,34 @@ local function fill_empty_hashes(hashes)
...
-    ngx_log(ngx_ERR, _log_prefix, "error while requesting version negotiation: " .. err)
-    assert(ngx.timer.at(math.random(5, 10), function(premature)
-      self:communicate(premature)
-    end))
-    return
+    ngx_log(ngx_ERR, _log_prefix, "error while requesting version negotiation: ", err)
+    ngx_log(ngx_ERR, _log_prefix, "Assuming old JSON protocol.")
+  end
+
+  local config_proto, msg = version_negotiation.get_negotiated_service("config")
```
Comment (suikabreaker) [inline kong/clustering/init.lua:220]: Should we still go though this if `request_version_handshake` fails? What do we get in that case?

Comment (suikabreaker) [reply kong/clustering/init.lua:220]: And if we already determined to use old protocol, we should not see the "error reading negotiated..." log.

File: kong/clustering/init.lua [hunk 2]
```diff
@@ -205,14 +205,34 @@ local function fill_empty_hashes(hashes)
...
+    self.child = require "kong.clustering.wrpc_data_plane".new(self)
+
+  elseif config_proto == "v0" or config_proto == nil then
+    self.child = require "kong.clustering.data_plane".new(self)
+  end
+
+  if self.child then
+    clustering_utils.load_config_cache(self.child)
+    self.child:communicate(premature, config_proto)
   end
```
Comment (suikabreaker) [inline kong/clustering/init.lua:236]: else?

File: kong/clustering/init.lua [hunk 3]
```diff
@@ -205,14 +205,34 @@ local function fill_empty_hashes(hashes)
...
 end
 
-function _M:request_version_negotiation()
+
+function _M:call_control_plane(premature)
```
Comment (suikabreaker) [inline kong/clustering/init.lua:209]: I think the function could be just "version\_negotiation" (though the name is already used).

Comment (Suika-Kong) [inline kong/clustering/init.lua:209]: The function's name could be simply "version\_negotiation" (though it's already used).

Comment (javierguerragiraldez) [reply kong/clustering/init.lua:209]: no, it does initiate the DP->CP connection.

File: kong/clustering/init.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/init.lua:236]: that would be a future "not nil, not v0, not v1" case.

**Comment by javierguerragiraldez** [comment]
as mentioned above, the version negotiation specs have changed (implemented in #8775 ), but the current master has at least one bug when failure to connect leads to restarting only part of the cycle, locking into the wrong protocol.  this patch fixes that by assuring that failures restart from the start.

**Pull request closed without merging**